### PR TITLE
Add CycleGAN structure and related processing 

### PR DESCRIPTION
**Pull request by Sebastian-Larsson**

### CODE REVIEW

File: models/cyclegan.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+        self.BPM_B = 120
+        self.counter = 0
+
+    def forward(self, x, gen_choice: str = 'A'):
+        if gen_choice == 'A':
+            return self.generator_A(x)
+        else:
+            return self.generator_B(x)
+
+    def training_step(self, batch: dict, batch_idx: dict, optimizer_idx: int):
```
Comment (LeifSulaiman) [inline models/cyclegan.py:95]: There is no need to define these input types. Remove types.

File: models/cyclegan.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+            return self.gen_step(segments_A, segments_B, 'A')
+        elif optimizer_idx == 1: # GEN
+            return self.gen_step(segments_A, segments_B, 'B')
+        elif optimizer_idx == 2: # DISC
+            return self.disc_step(segments_A, segments_B, 'A')
+        elif optimizer_idx == 3: # DISC
+            return self.disc_step(segments_A, segments_B, 'B')
+
+    def gen_step(self, segments_A, segments_B, model_choice: str = 'A'):
+        generator_loss = 0
```
Comment (LeifSulaiman) [inline models/cyclegan.py:116]: Remove this line and return generator loss in each if statement instead.

File: models/cyclegan.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+            loss_cycle_B = self.cycle_loss(rec_B, segments_B)
+            generator_loss = loss_modifier_B + loss_cycle_B * self.hparams.lmd
+
+            self.log("Cycle loss B", loss_cycle_B, on_step=False, on_epoch=True, prog_bar=False, logger=True)
+            self.log("GAN loss B", loss_modifier_B, on_step=False, on_epoch=True, prog_bar=False, logger=True)
+
+        return generator_loss
+
+    def disc_step(self, segments_A, segments_B, model_choice: str = 'A'):
+        disc_loss = 0
```
Comment (LeifSulaiman) [inline models/cyclegan.py:146]: Same here, remove this initialization and return loss in each if.

File: models/cyclegan.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+                    directory=self.hparams.save_directory,
+                    bpm=self.BPM_B,
+                    resolution=self.hparams.resolution)
+
+    def adversarial_loss(self, y_hat, y):
+        return F.binary_cross_entropy(y_hat, y)
+
+    def init_weights(self, model):
+        for m in model.modules():
+            if isinstance(m, (nn.Conv2d, nn.ConvTranspose2d)):
```
Comment (LeifSulaiman) [inline models/cyclegan.py:250]: Note that the bias is not initialized for convolutional layers. Is this intended? Some discriminators/generators in the commit have bias in their convolutional layers.

File: models/example_train_cyclegan.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+                         resolution=8,
+                         gen_num_channels=32,
+                         disc_num_channels=16,
+                         input_ch_size=14)
+
+    # Create trainer then train the model
+    print('\nLogs are stored in "{}".'.format(LOG_PATH))
+    trainer = pl.Trainer(
+            max_epochs=EPOCHS,
+            gpus=1,
```
Comment (LeifSulaiman) [inline models/example_train_cyclegan.py:40]: Please refrain from writing gpu= some int. I have made a similar mistake. It is more optimal to define a torch.device.

File: models/generators.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        conv6 = self.conv6(conv5_pool)
+
+        # Upsample
+        upconv5 = self.upconv5(conv6)
+        upconv4 = self.upconv4(torch.cat([upconv5, conv5], 1))
+        upconv3 = self.upconv3(torch.cat([upconv4, conv4], 1))
+        upconv2 = self.upconv2(torch.cat([upconv3, conv3], 1))
+        upconv1 = self.upconv1(torch.cat([upconv2, conv2], 1))
+        output = self.output(self.sideconv1(torch.cat([upconv1, conv1], 1)))
+        return torch.clamp(output, min=0, max=127)
```
Comment (LeifSulaiman) [inline models/generators.py:71]: I don't know if this an optimal approach. This is basicly using relu as output activation. Doesn't it make more sense to clamp when not training, i.e. on test data and even in samples in the pl.lightning module.

File: models/generators.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+    def Encoder(self, in_channels, out_channels, kernel_size, padding, batch=True):
```
Comment (LeifSulaiman) [inline models/generators.py:73]: Rename argument 'batch' to 'batchnorm' or something similar. Same for 'Decoder' function.

File: models/generators.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+            models.resnet.BasicBlock(num_channels*4, num_channels*4),
+            self._deconv_block(num_channels*4, num_channels*2, 3, 2, 1, 1),
+            self._deconv_block(num_channels*2, num_channels, 3, 2, 1, 1),
+            self._deconv_block(num_channels, input_ch_size, 7, 1, 3),
+        )
+
+    def forward(self, x):
+        return torch.clamp(self.network(x), min=0, max=127)
+
+    def _conv_block(self, input_kernels, output_kernels, kernel_size=7, stride=1, padding=0):
```
Comment (LeifSulaiman) [inline models/generators.py:135]: For consistency, call the functions 'Encoder' and 'Decoder' instead of 'conv_block' and 'deconv_block'.

File: processing/npy_to_midi_example.py [hunk 9]
```diff
@@ -5,14 +5,15 @@
...
+test_song = song_array[0]
+bpm = song_array[1]
+resolution = 8 # Preferably same as used in preprocessing.
+
 def main():
-    print(bpm)
     # Normally you'd put this at the end of an epoch in training and as such you wouldn't have to convert it
     # into a tensor.
-    song_tensor = torch.tensor(test_song[0][:-1]) # Exclude last layer, it contains the bpms
+    song_tensor = torch.tensor(test_song[0]) # Exclude last layer, it contains the bpms
```
Comment (LeifSulaiman) [inline processing/npy_to_midi_example.py:16]: Remove comment as it doesn't apply anymore.

File: models/generators.py (hunk 6 above)
Comment (Sebastian-Larsson) [reply models/generators.py:71]: Changing so it only clamps when not training.

File: models/cyclegan.py (hunk 4 above)
Comment (Sebastian-Larsson) [reply models/cyclegan.py:250]: The current implementation does not need bias init.

**Review by LeifSulaiman — APPROVED** [review]
Approve changes.
